### PR TITLE
Add omega scans to workflow

### PR DESCRIPTION
**Pull request by spxiwh**
This patch adds an omega scans executable to PyCBC and implements it within the PyCBC workflow. An example of this is here:

https://ldas-jobs.ligo.caltech.edu/~spxiwh/aLIGO/O2/analyses/O2_chunk19_HM/run3B/3._single_triggers/3.07_H1_loudest_noncoinc_time/

These new plots will appear in injection, single and coincident minifollowups.

This requires some minor changes to configuration files to add the new executable and options. Please advise on what file I should add those changes to?

### CODE REVIEW

**Comment by tdent** [comment]
ok, on the colorbar: i suggest a default of log scaling and displaying one colorbar with log10 ticks, with a range up to the maximum over subplots, and I'd be OK with different subplots all using the same color scale.  (if you can persuade matplotlib to make a single colorbar beside 3 subplots ..)
I think this will work _because_ the log scaling should be able to show both quieter and louder glitches/signals at the same time.  And if there _is_ a louder glitch several seconds away from trigger merger time we'd want to know about it.

**Comment by spxiwh** [comment]
@tdent Updated example to address your comment above:

https://ldas-jobs.ligo.caltech.edu/~spxiwh/H1-PLOT_QSCAN_2-1185217218-720000.png

What should the colorbar label say?

Adding the colorbar moves the x-axis label a little off-center. I think it's going to be a lot more effort to figure out how to resolve that than it's worth!

**Comment by tdent** [comment]
'Normalized power' or 'Normalized energy' ? 
Off-center x label is fine IMO since most of the subplots arent centered at 0 anyway! 

BTW, would be nice to see the standard set of time windows around the 170817 trigger in each ifo as a test ..

**Comment by spxiwh** [comment]
Hmmm:

https://ldas-jobs.ligo.caltech.edu/~spxiwh/H1-PLOT_QSCAN_2-1185217218-720000.png

https://ldas-jobs.ligo.caltech.edu/~spxiwh/L1-PLOT_QSCAN_2-1185217218-720000.png

**Comment by tdent** [comment]
ah .. the qrange is totally wrong for BNS except for the lowest H1 plot. 
If you force q to be 50 or larger it will look 'good' but not sure what one can do about that in general.

File: bin/plotting/pycbc_plot_qscan [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                         'Optional, default=10')
+parser.add_argument('--qtransform-frange-upper', default=None, type=float,
+                    help='Upper frequency at which to compute qtransform. '
+                         'Optional, default=Half of Nyquist')
+parser.add_argument('--qtransform-qrange-lower', default=4, type=float,
+                    help='Lower limit of the range of q to consider, '
+                         'default=4')
+parser.add_argument('--qtransform-qrange-upper', default=64, type=float,
+                    help='Upper limit of the range of q to consider, '
+                         'default=4')
```
Comment (tdent) [inline bin/plotting/pycbc_plot_qscan]: presumably default=4 is a typo

File: bin/plotting/pycbc_plot_qscan [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        return [start, end]
+    except:
+        raise argparse.ArgumentTypeError("Input must be start,end start,end")
+
+parser = argparse.ArgumentParser(description=__doc__)
+parser.add_argument("--version", action="version",
+                    version=pycbc.version.git_verbose_msg)
+parser.add_argument('--output-file', required=True, help='Output plot')
+parser.add_argument('--center-time', type=float,
+                    help='Center plot on the given GPS time')
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_qscan]: Is `--center-time` not a required argument?

Comment (titodalcanton) [reply bin/plotting/pycbc_plot_qscan]: Nevermind, it makes sense after reading the other options.

File: bin/plotting/pycbc_plot_qscan [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    ax.set_ylim(curr_frange[0], curr_frange[1])
+    if opts.log_y:
+        ax.set_yscale('log')
+
+# https://stackoverflow.com/questions/6963035/pyplot-axes-labels-for-subplots
+fig.add_subplot(111, frameon=False)
+plt.tick_params(labelcolor='none', top='off', bottom='off', left='off',
+                right='off')
+plt.grid(False)
+plt.xlabel('Time from {} (s)'.format(opts.center_time))
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_qscan]: I would use `{:.3f}` here to make sure we consistently get millisecond precision.

File: bin/plotting/pycbc_plot_qscan [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+plt.ylabel('Frequency (Hz)')
+
+# https://stackoverflow.com/questions/13784201/matplotlib-2-subplots-1-colorbar
+cb = fig.colorbar(im, ax=axes.ravel().tolist())
+cb.set_label('What should this say?')
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_qscan]: "Normalized energy", "Normalized signal energy", "Signal energy [au]"…

File: bin/plotting/pycbc_plot_qscan [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+if opts.plot_title is None:
+    opts.plot_title = 'Q-transform plot'
+if opts.plot_caption is None:
+    # FIXME: Someone please improve!
+    opts.plot_caption = ("This shows the Q-transform as a function of time and "
+                        "frequency")
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_qscan]: I would also store the center time in the default title and/or caption for reference, as written in the x label.

**Comment by titodalcanton** [comment]
As for the config file to add these options to, maybe the O3 development file?

https://github.com/ligo-cbc/pycbc-config/tree/master/O3/pipeline

**Comment by spxiwh** [comment]
Indeed setting the lower qrange to 50 results in better plots:

https://ldas-jobs.ligo.caltech.edu/~spxiwh/H1-PLOT_QSCAN_3-1185217218-720000.png

https://ldas-jobs.ligo.caltech.edu/~spxiwh/L1-PLOT_QSCAN_3-1185217218-720000.png

File: bin/plotting/pycbc_plot_qscan (hunk 1 above)
Comment (spxiwh) [reply bin/plotting/pycbc_plot_qscan]: Fixed

File: bin/plotting/pycbc_plot_qscan (hunk 2 above)
Comment (spxiwh) [reply bin/plotting/pycbc_plot_qscan]: Added required=True here.

File: bin/plotting/pycbc_plot_qscan (hunk 3 above)
Comment (spxiwh) [reply bin/plotting/pycbc_plot_qscan]: Done

File: bin/plotting/pycbc_plot_qscan (hunk 4 above)
Comment (spxiwh) [reply bin/plotting/pycbc_plot_qscan]: Changed as suggested by Tom above

File: bin/plotting/pycbc_plot_qscan (hunk 5 above)
Comment (spxiwh) [reply bin/plotting/pycbc_plot_qscan]: Changed

**Comment by spxiwh** [comment]
Okay, I think this is ready and I've addressed all comments now. I'll do one more quick test run and then look to push this with the requested config file changes.

**Comment by spxiwh** [comment]
What's the benefit to a lower q-value than 50?

**Comment by ahnitz** [comment]
I haven't quite followed how you are choosing q for each of the panes. Is that a separate choice? If  you are setting a blanket minimum at 50, you won't be able to effectively resolve any time-frequency evolution of BBH signals.

**Comment by spxiwh** [comment]
Hi Alex,

I'm just following what timeseries.gtransform does. E.g. the "q" values are allowed to range between 4 and 64 (by default, which one can override on the command line, which gives the kwarg to timeseries.qtransform).

**Comment by spxiwh** [comment]
https://github.com/ligo-cbc/pycbc-config/pull/27

**Pull request merged**